### DISCUSSION TITLE
Learnable prior over parameters

### DISCUSSION THREAD

**Discussion by fweberling** [cornellius-gp/gpytorch#1731] [Q&A] [answered]
Hi everyone,

I have a kernel with a parameter to which I assigned a prior distribution. I would now like to learn the parameters of this prior distribution. 

Is that possible in GPyTorch? 

So far, I have only seen that I can assign a prior distribution with fixed but not learnable parameters.

Does somebody might know a way around this issue?

Thanks a lot!

**Comment by jacobrgardner** [comment, marked as answer]
I mean, purely mechanically, this should be possible by just declaring the hyperparameters of a prior you define extending one of our existing ones to be true `torch.nn.Parameter`, since our priors all extend `Module`. Currently, they are made buffers via: 
https://github.com/cornellius-gp/gpytorch/blob/98dd6169e4ea0a9974498ed65b8070b9d8ed824f/gpytorch/priors/utils.py#L4
e.g.:
https://github.com/cornellius-gp/gpytorch/blob/98dd6169e4ea0a9974498ed65b8070b9d8ed824f/gpytorch/priors/torch_priors.py#L74

You would just use `register_parameter` instead of `register_buffer`.

I'm not sure that this is a good idea though. When doing MAP, it seems like the optimizer would just set the prior parameters such that whatever parameter values maximize the marginal likelihood would maximize the prior also. It's not obvious to me that running NUTS would work at all.
